src/game/common/rts/playerlist.cpp

```diff
@@ -18,3 +18,234 @@
+    }
+    return nullptr;
+}
+
+// zh: 0x0045B0A0 wb: 0x007BF613
+void PlayerList::Team_About_To_Be_Deleted(Team *team)
+{
+    for (auto *player : m_players) {
+        player->Remove_Team_Relationship(team);
+    }
+}
+
+// zh: 0x0045B0D0 wb: 0x007BF64F
+void PlayerList::Update_Team_States()
+{
+    for (auto *player : m_players) {
+        player->Update_Team_States();
+    }
+}
+
+// zh: 0x0045B0F0 wb: 0x007BF685
+Team *PlayerList::Validate_Team(Utf8String name)
+{
+#ifdef GAME_DLL
+    return Call_Method<Team *, PlayerList, Utf8String>(PICK_ADDRESS(0x0045B0F0, 0x007BF685), this, name);
+#else
+    // Requires TeamFactory::Find_Team
+    return nullptr;
+#endif
+}
+
+// zh: 0x0045B170 wb: 0x007BF739
+void PlayerList::Set_Local_Player(Player *player)
+{
+    if (player == nullptr) {
+        captainslog_dbgassert(false, "Local player may not be null");
+        player = Get_Neutral_Player();
+    }
+
+    if (player == m_local) {
+        return;
+    }
+
+    if (m_local != nullptr) {
+        m_local->Becoming_Local_Player(false);
+    }
+    m_local = player;
+    player->Becoming_Local_Player(true);
+}
+
+// zh: 0x0045B1B0 wb: 0x007BF7BD
+Player *PlayerList::Get_Player_From_Mask(uint16_t mask)
+{
+    for (auto i = 0; i < MAX_PLAYER_COUNT; ++i) {
+        auto *player = Get_Nth_Player(i);
+        if (player == nullptr) {
+            continue;
+        }
+
+        if (player->Get_Player_Mask() & mask) {
+            return player;
+        }
+    }
+    captainslog_dbgassert(false, "Player does not exist for mask");
+    return nullptr;
+}
+
+// zh: 0x0045B1F0 wb: 0x007BF855
+Player *PlayerList::Get_Each_Player_From_Mask(uint16_t &mask)
+{
+    for (auto i = 0; i < MAX_PLAYER_COUNT; ++i) {
+        auto *player = Get_Nth_Player(i);
+        if (player == nullptr) {
+            continue;
+        }
+
+        if (player->Get_Player_Mask() & mask) {
+            mask &= ~player->Get_Player_Mask();
+            return player;
+        }
+    }
+    captainslog_dbgassert(false, "No players found that contain any matching masks.");
+    mask = 0;
+    return nullptr;
+}
+
+// zh: 0x0045B250 wb: 0x007BF914
+uint16_t PlayerList::Get_Players_With_Relationship(int32_t player_idx, uint16_t relationship_mask)
+{
+    if (relationship_mask == 0) {
+        return 0;
+    }
+
+    uint16_t player_mask = 0;
+    auto *target_player = Get_Nth_Player(player_idx);
+    if (target_player == nullptr) {
+        return 0;
+    }
+
+    if (relationship_mask & (1ULL << 0)) {
+        player_mask |= target_player->Get_Player_Mask();
+    }
+
+    for (auto i = 0; i < Get_Player_Count(); ++i) {
+        auto *player = Get_Nth_Player(i);
+        if (player == nullptr || player == target_player) {
+            continue;
+        }
+
+        const auto relationship = target_player->Get_Relationship(player->Get_Default_Team());
+        switch (relationship) {
+            case Relationship::ENEMIES:
+                if (relationship_mask & (1ULL << 2)) {
```

Comment:
why is this more preferable than just & 4?

Reply:
These are flags. I was going to switch them to an enum but forgot. Will add that shortly.